# Solución de sistemas lineales de ecuaciones algebráicas

## Motivación

Los sistemas lineales de ecuaciones algebráicas son útiles para modelar muchos problemas científicos tales como el balanceo de ecuaciones químicas, el cálculo de corrientes en circuitos electrónicos, además de tener muchísimas aplicaciones dentro de la computación y las mismas matemáticas. Por ello, es de gran utilidad saber cómo se pueden implementar métodos numéricos que resuelvan este tipo de sistemas.

## Sistemas lineales de ecuaciones algebráicas

Un _sistema de ecuaciones_ es un grupo de ecuaciones que deben cumplirse simultáneamente. En general, podemos expresar un sistema de $m$ ecuaciones y $n$ incógnitas como

$$f_1(x_1,x_2,\dots,x_n) = 0,$$

$$f_2(x_1,x_2,\dots,x_n) = 0,$$

$$\vdots$$

$$f_m(x_1,x_2,\dots,x_n) = 0,$$

donde $f_j$ es una función arbitraria, con $1\leq j\leq m$. Una _solución_ al sistema es una $n$-tupla de valores $(x_1,x_2,\dots,x_n)$ para los cuales todas las ecuaciones del sistema se verifican. Un _sistema de ecuaciones algebráicas_ es aquel donde todas las funciones $f_j$ son **polinomios** en las variables $x_1,x_2,\dots,x_n$; más aún, un sistema de este tipo es _lineal_ si en todas las ecuaciones del sistema las incógnitas $x_1,x_2,\dots,x_n$ **no aparecen elevadas a una potencia mayor que uno** y **tampoco aparecen productos entre las diferentes incógnitas**.

En este _notebook_ veremos un método para solucionar sistemas lineales de ecuaciones algebráicas de $n$ ecuaciones y $n$ variables, donde todas las ecuaciones son _linealmente independientes_. En general, este tipo de sistemas se pueden escribir de la siguiente forma (verifica que cumpla las definiciones anteriores):

$$ a_{11}x_1 + a_{12}x_2 + \dots + a_{1n}x_n = b_1, $$

$$ a_{21}x_1 + a_{22}x_2 + \dots + a_{2n}x_n = b_2, $$

$$ \vdots $$

$$ a_{n1}x_1 + a_{n2}x_2 + \dots + a_{nn}x_n = b_n. $$

### Operaciones elementales

Existen tres "operaciones elementales" que le podemos aplicar a este sistema **sin cambiar su conjunto de soluciones**:
1. multiplicar una ecuación (de ambos lados) por una constante no nula;
1. sumarle a una ecuación alguna de las otras ecuaciones multiplicada por una constante arbitraria;
1. intercambiar dos ecuaciones de lugar.

El hecho de que la operación 1 no cambie al conjunto de soluciones del sistema se sigue de que, para todo $1\leq i\leq n$, si $c\neq0$, entonces

$$c(a_{i1}x_1 + a_{i2}x_2 + \dots + a_{in}x_n - b_i) = 0 \iff a_{i1}x_1 + a_{i2}x_2 + \dots + a_{in}x_n - b_i = 0.$$

Similarmente, el hecho de que la operación 2 no cambie al conjunto de soluciones del _sistema_ se sigue de que, para cualesquiera $1\leq i,j\leq n$ y para cualquier constante arbitraria $c'$,

\begin{align}
    (a_{i1}x_1 + a_{i2}x_2 + \dots + a_{in}x_n - b_i) = 0 \ &\land \ (a_{j1}x_1 + a_{j2}x_2 + \dots + a_{jn}x_n - b_j) = 0 \\
    &\iff \\
    (a_{i1}x_1 + a_{i2}x_2 + \dots + a_{in}x_n - b_i) \ &+ \ c'(a_{j1}x_1 + a_{j2}x_2 + \dots + a_{jn}x_n - b_j) = 0.
\end{align}

En efecto: La primera implicación es directa, mientras que la otra se sigue de que las ecuaciones son linealmente dependientes. Por último, es claro por qué la operación 3 no cambia el conjunto de soluciones pues, después de aplicarla, las ecuaciones del sistema siguen siendo las mismas.

## Método de eliminación Gaussiana

A continuación, veremos un algoritmo que aprovecha las operaciones elementales descritas anteriormente para obtener una solución a un sistema lineal de ecuaciones algebráicas.

### Retropropagación

Supongamos que, a través de operaciones elementales, podemos transformar el sistema lineal de ecuaciones algebráicas anterior en uno de la forma

\begin{align*}
    a'_{11}x_1 \ + \ a'_{12}x_2 \ + \dots + \ a'_{1n-1}x_{n-1} \ + \ a'_{1n}x_n &= b'_1, \\
    0 \ + a'_{22}x_2 + \dots + a'_{1n-1}x_{n-1} \ + \ a'_{1n}x_n &= b'_2, \\
    & \ \vdots \\
    0 + 0 + \dots + a'_{n-1,n-1} x_{n-1} + a'_{n-1n}x_n &= b'_{n-1}, \\
    0 \ + \ 0 \ + \ \dots \ + \ \ 0 \ \ + \ \ \ a'_{nn}x_n &= b'_n,
\end{align*}

con $a'_{ii}\neq0$ para $1\leq 1\leq n$ (en la sección **Reducción** explicaremos por qué es posible lograr esto). Observemos que la última ecuación tiene sólo una incógnita, que podemos despejar para obtener

$$x_n = \frac{b'_n}{a'_{nn}}.$$

**Ejercicio** Despeja a $x_{n-1}$ de la penúltima ecuación y sustituye en valor de $x_n$.

De la ecuación $$ a'_{n-1,n-1} x_{n-1} = b'_{n-1} - a'_{n-1n}x_n $$ obtenemos que

$$x_{n-1} = \frac{(b'_{n-1} -  a'_{n-1n})x_n}{a'_{n-1,n-1}}$$

Ahora sustituimos el valor $x_n$

$$x_{n-1} = \frac{(b'_{n-1} -  a'_{n-1n})b'_n}{a'_{nn}a'_{n-1,n-1}}$$

Iterativamente, por cada ecuación que solucionamos, podemos sustituir todos los valores obtenidos hasta el momento en la ecuación _anterior_ para solucionarla también, hasta haber encontrado valores $(x_1,x_2,\dots,x_n)$ que solucionen todo el sistema. A este procedimiento se le conoce como _retropropagación_.

**Ejercicio** Da una fórmula general para el valor de $x_i$, con $1\leq i<n$, en términos de los $x_j$ para $i < j \leq n$.

Para algún $x_i$, con $1\leq i<n$ 
$$x_i= \frac{b'_{i-1}-\sum_{j-i+1}^{n}a'_{i-1}x_{j}}{a_{ii}} $$

**Nota** Lo que nos asegura que los coeficientes $a_{ii}$ con $1\leq i\leq n$ son distintos de cero y, por lo tanto, que nuestras soluciones para $x_n, x_{n-1},\dots,x_1$ están bien definidas es que las ecuaciones del sistema sean linealmente independientes. Esto es equivalente a que la matriz cuadrada tenga un determinante distinto de cero. A este tipo de matrices se les llama _matrices no singulares_.

### Reducción

Al proceso de tomar un sistema lineal de $n$ ecuaciones algebráicas y $n$ incógnitas

$$ a_{11}x_1 + a_{12}x_2 + \dots + a_{1n}x_n = b_1, $$

$$ a_{21}x_1 + a_{22}x_2 + \dots + a_{2n}x_n = b_2, $$

$$ \vdots $$

$$ a_{n1}x_1 + a_{n2}x_2 + \dots + a_{nn}x_n = b_n, $$

y transformarlo a través de operaciones elementales en uno de la forma

\begin{align*}
    a'_{11}x_1 \ + \ a'_{12}x_2 \ + \dots + \ a'_{1n-1}x_{n-1} \ + \ a'_{1n}x_n &= b'_1, \\
    0 \ + a'_{22}x_2 + \dots + a'_{1n-1}x_{n-1} \ + \ a'_{1n}x_n &= b'_2, \\
    & \ \vdots \\
    0 + 0 + \dots + a'_{n-1,n-1} x_{n-1} + a'_{n-1n}x_n &= b'_{n-1}, \\
    0 \ + \ 0 \ + \ \dots \ + \ \ 0 \ \ + \ \ \ a'_{nn}x_n &= b'_n.
\end{align*}

se le conoce como _reducción_, y esto último se conoce como la _forma reducida_ del sistema. Una _transformación elemental_ es la transformación de una ecuación del sistema en otra a través de operaciones elementales.

Sea $E_i$ la $i$-ésima ecuación del sistema no reducido. Si $E_1$ tiene un coeficiente $a_{11}$ distinto de cero, definimos

$$ E_1 =: E'_1; \quad (1) $$

de lo contrario, primero intercambiemos a $E_1$ por alguna ecuación $E_j$ con el coeficiente $a_{j1}$ de $x_1$ distinto de cero

$$E_1 \leftrightarrow E_j$$

y después definimos a $E_1'$ usando la ecuación (1). En ambos casos, redefinamos sus coeficientes como $a'_{11}:= a_{j1}, a'_{12}:= a_{j2},\dots, b'_1:=b_j$, obteniendo la ecuación

$$E'_1: a'_{11}x_1 + a'_{12}x_2 + ... + a'_{1n} x_n = b'_1,$$

con $a'_{11}\neq0$.

Ahora, observemos que, si $E_2$ tiene un coeficiente $a_{22}$ distinto de cero, entonces al aplicar la transformación elemental

$$E_2 \to E_2 - \frac{a_{21}}{a'_{11}} E_1' =: E'_2 \quad (2)$$

la ecuación resultante $E'_2$ tendrá un coeficiente nulo en $x_1$ y no nulo en $x_2$ (Escríbela a mano y verifícalo); en cambio, si $a_{22} = 0$, primero debemos intercambiar a $E_2$ por alguna ecuación $E_k$ con el coeficiente $a_{k2}$ de $x_2$ distinto de cero

$$ E_2 \leftrightarrow E_k $$

y después definir a $E'_2$ con la ecuación (2) para que la ecuación resultante tenga un coeficiente nulo en $x_1$ y no nulo en $x_2$. En ambos casos, definimos a los coeficientes de la ecuación $E'_2$, obteniendo

$$E'_2: a'_{21}x_1 + a'_{22}x_2 + ... + a'_{2n} x_n = b'_2,$$

con $a'_{21}=0$ y $a'_{22}\neq0$.

Luego, si $E_3$ tiene un coeficiente $a_{33}$ distinto de cero, entonces al aplicar la transformación elemental

$$ E_3 \to E_3 - \frac{a_{31}}{a'_{11}} E_1' - \frac{a_{32}}{a'_{22}} E_2' =: E'_3 \quad (3) $$

la ecuación resultante $E'_3$ tendrá coeficientes nulos en $x_1$ y $x_2$, y un coeficiente no nulo en $x_3$; en cambio, si $a_{33} = 0$, primero debemos intercambiar a $E_3$ por alguna ecuación $E_m$ con el coeficiente $a_{l3}$ de $x_3$ distinto de cero

$$ E_3 \leftrightarrow E_l $$

y después definir

y después definir a $E'_3$ con la ecuación (3) para que la ecuación resultante tenga un coeficiente nulo en $x_1$ y $x_2$, y no nulo en $x_3$. En ambos casos, definimos a los coeficientes de la ecuación $E'_3$, obteniendo

$$E'_3: a'_{31}x_1 + a'_{32}x_2 + ... + a'_{3n} x_n = b'_3,$$

con $a'_{31}=a'_{32}=0$ y $a'_{33}\neq0$. (¿Observas un patrón? En serio, es muy importante que escribas al menos estos dos primeros pasos a mano, para que realmente lo _observes_.).

Podemos seguir aplicando el mismo procedimiento iterativamente hasta haber obtenido ecuaciones $E'_1, E'_2, \dots, E'_n$, las cuales forman un sistema _reducido_ al cual le podemos aplicar _retropropagación_.

**Nota** La razón por la que **debe existir** una ecuación $E_j$ con coeficiente $a_{j1}\neq0$ con el cual podamos definir a $E'_1$ y empezar el proceso de reducción es que, de lo contrario, ¡el sistema tendría $n-1$ incógnitas en vez de $n$!

Observemos que las transformaciones elementales **no afectan a las incógnitas** $x_1,x_2\dots,x_n$, sino que **sólo afectan a los coeficientes** $a_{ij}, b_i$, con $1\leq i,j\leq n$. Por lo tanto, podemos hacer reducción _sólo considerando a los coeficientes del sistema_. Similarmente, notemos que, para hacer retropropagación, _sólo necesitamos a los coeficientes del sistema reducido_. Esto nos lleva desechar los sistemas en favor de una representación más sencilla de visualizar y de implementar en código.

### Representación matricial


Consideremos, nuevamente, un sistema lineal de ecuaciones algebráicas

$$ a_{11}x_1 + a_{12}x_2 + \dots + a_{1n}x_n = b_1, $$

$$ a_{21}x_1 + a_{22}x_2 + \dots + a_{2n}x_n = b_2, $$

$$ \vdots $$

$$ a_{n1}x_1 + a_{n2}x_2 + \dots + a_{nn}x_n = b_n. $$

Como tal vez sospeches por la notación de los coeficientes, podemos escribir este sistema como una sola ecuación vectorial

$$ \begin{pmatrix} a_{11} &a_{12} &\dots &a_{1n} \\ a_{21} &a_{22} &\dots &a_{2n} \\ \vdots &\vdots &\ddots &\vdots \\ a_{n1} &a_{n2} &\dots &a_{nn} \end{pmatrix} \begin{pmatrix} x_1 \\ x_2 \\ \vdots \\ x_n \end{pmatrix} = \begin{pmatrix} b_1 \\ b_2 \\ \vdots \\ b_n \end{pmatrix}; $$

o, de forma más sucinta,

$$ A \vec{x} = \vec{b}. $$

Observemos que podemos representar a nuestro sistema de ecuaciones con la matriz aumentada

$$ \big( A\mid\vec{b} \ \big) = \begin{pmatrix} a_{11} &a_{12} &\dots &a_{1n} &\mid &b_1 \\ a_{21} &a_{22} &\dots &a_{2n} &\mid &b_2 \\ \vdots &\vdots &\ddots &\vdots &\mid &\vdots \\ a_{n1} &a_{n2} &\dots &a_{nn} &\mid &b_n \end{pmatrix}. $$

De esta forma, cada renglón de la matriz aumentada representa una ecuación de nuestro sistema... ¡y las operaciones elementales que vimos para sistemas lineales de ecuaciones _coinciden completamente_ con operaciones elementales de matrices! (De hecho, de ahí viene el nombre.) Recordando que las transformaciones elementales sólo afectan a los **coeficientes** de un sistema lineal de ecuaciones, entonces podemos aplicarle **reducción** a la matriz aumentada para llevarla a su forma reducida

$$ \begin{pmatrix} a'_{11} &a'_{12} &\dots &a'_{1n} &\mid &b'_1 \\ 0 &a'_{22} &\dots &a'_{2n} &\mid &b'_2 \\ \vdots &\vdots &\ddots &\vdots &\mid &\vdots \\ 0 &0 &\dots &a'_{nn} &\mid &b'_n \end{pmatrix}. $$

Luego, recordando que la retropropagación se puede llevar a cabo sólo con los coeficientes del sistema reducido, ¡podemos utilizar esta matriz reducida para encontrar la solución de nuestro sistema!

## Implementación

Para este punto, ya debes tener motivación para resolver un problema importante y ubicuo, así como un plan de acción respaldado por una teoría sensata. Ahora sí, ¡te toca implementarlo!

Como sugerencia general, te recomiendo recordar que puedes crear celdas de Jupyter libremente para experimentar con tu código, probarlo con diferentes parámetros, ver si funciona como esperas, e irlo corrigiendo en caso de que no. Además, si así lo deseas, puedes empezar resolviendo sistemas pequeños (por ejemplo, de dos ecuaciones) y, ya que tengas una noción de los procedimientos y cómo implementarlos, generalizar tu código para que también pueda resolver sistemas más grandes. 

**Ejercicio** Crea una función `retropropagación` con un argumento `M`, que puedes suponer que es una matriz aumentada de la forma

$$ \begin{pmatrix} a'_{11} &a'_{12} &\dots &a'_{1n} &\mid &b'_1 \\ 0 &a'_{22} &\dots &a'_{2n} &\mid &b'_2 \\ \vdots &\vdots &\ddots &\vdots &\mid &\vdots \\ 0 &0 &\dots &a'_{nn} &\mid &b'_n \end{pmatrix} $$

-es decir, aquella formada por una matriz triangulada superior y un vector $\vec{b}'$-, que aplique retropropagación a la matriz `M` y devuelva como resultado un arreglo con el vector solución $\vec{x}$.

In [3]:
function retropropagación(M,b)                           # Definimos la función retropropagación
    n = size(M,1)                                        # definimos n como el tamaño de la matriz
    x = zeros(n)                                         # definimos x como ceros para la
    x[n] = b[n]/M[n,n]                                   # x[n] lo definimos como b/M
    for i in n-1:-1:1                                    # para i en n-1 a 1
        s = sum( M[i,j]*x[j] for j in i+1:n )            # s es la suma de M*x para j en i+1 a n
        x[i] = ( b[i] - s ) / M[i,i]                     # xi lo definimos como bi - la suma s entre la matriz
                                                         # este proceso se repite hasta n
    end
    return x                                             # regresamos x que es el vector solución
end

retropropagación (generic function with 1 method)

In [9]:
A = [3 4 7; 9 5 3; 2 5 9]
b = [1 2 3]
retropropagación(A,b)

3-element Vector{Float64}:
 -0.7111111111111109
  0.2
  0.3333333333333333

Ya hemos implementado un algoritmo capaz de solucionar matrices reducidas. Ahora, implementaremos otro que realice el proceso de reducción de una matriz aumentada. Como este proceso depende de que las entradas que van quedando en la diagonal sean no nulas -para poder dividir entre ellas-, primero implementaremos una función que se asegure de que esto suceda.

**Ejercicio** Crea una función `pivote` que tome tres argumentos `M`, `i` y `j`, y revise si la entrada del renglón `i` y columna `j` de la matriz `M` es igual a cero. En caso que sí lo sea, debe buscar la primera entrada distinta de cero entre los siguientes renglones de la misma columna y, al encontrarla, intercambiar el renglón correspondiente a la entrada no nula con el renglón `i`.

Sugerencias: Recuerda que puedes aplicar la función `length` a un vector columna de una matriz para obtener la cantidad de renglones de esa matriz. Además, puedes crear una variable temporal para no perder información durante el intercambio de los renglones.

In [37]:
function pivote(M,i,j,b)                              # Definimos la función pivote
    n,m = size(M)                                     # definimos n y m como el tamaño de la matriz
    n_b = length(b)                                   # y n_b el tamaño del vector b
    
    i_max = i                                        # definimos i maximo
    for k = i-1:n                                    # para una k igual a i-1 hasta n
        if abs(M[k,j]) > abs(M[i_max, j])           # si el valor absoluto de M de k,j es mayor al valor absoluto de M de i_max,j
            i_max = k                               # definimos i_max como k
        end
    end
     if i_max != i                                     # si el i_max es distinto de i
        M[i,:], M[i_max,:] = M[i_max,:], M[i,:]        # definimos M para i, i_max
        b[i], b[i_max] = b[i_max], b[i]                # definimos b para i, i_max
    end
     return M,b                                       # regresamos a la matriz y el vector b
end

pivote (generic function with 1 method)

In [39]:
A = [3 4 7; 9 5 3; 2 5 9]
b = [1 2 3]
pivote(A,2,3,b)

([3 4 7; 2 5 9; 9 5 3], [1 3 2])

**Ejercicio** Crea una función `reducción` que tome un argumento `M`, que puedes suponer que es una matriz aumentada de la forma

$$ \begin{pmatrix} a_{11} &a_{12} &\dots &a_{1n} &\mid &b_1 \\ a_{21} &a_{22} &\dots &a_{2n} &\mid &b_2 \\ \vdots &\vdots &\ddots &\vdots &\mid &\vdots \\ a_{n1} &a_{n2} &\dots &a_{nn} &\mid &b_n \end{pmatrix} $$

que representa un sistema de ecuaciones algebráicas linealmente independientes -y que, por ende, le podemos aplicar reduccion-, que devuelva la matriz reducida correspondiente.

Sugerencia: Usa la función `pivote` que creaste anteriormente.

In [28]:
function reduccion(M,b)                                           # nombramos a la función reduccion
    n,m = size(M)                                                 # c y m nos dara el tamaño de la matriz
    n_b = length(b)                                               # nos dara el tamaño del vector b
    if n != m                                                     # si n es distinta de m sale un mensaje de error
        error("La matriz no es cuadrada ")
    elseif n_b != n                                               # si el tamaño del vector es distinto a n sale un mensaje de error
        error("La longitud del vector de soluciones debe ser igual a las incognitas")
    end
    
    MB = hcat(M,B)                                               # definimos MB como la concatenacion de m y b
    for j =1:n                                                   # para j de 1 a n
        i_max = j                                                # definimos i maxima
        for i = j+1:n                                            # definimos i como j+1 hasta n
            if abs(MB[i,j]) > abs(MB[i_max,j])                   # si el valor absoluto de mb de i a j es mayor que el valor absoluto de mb de imaxima a j
                i_max = i                                        # definimos i maxima como i
            end
        end
        
        if i_max != j                                            # si el i maximo es distinto de j
            MB[j,:], MB[i_max,:] = MB[i_max,:], MB[j,:]          
        end
        
        for i= j+1:n                                             # para i igual a j+1 hasta n
            c = MB[i,j]/MB[j,j]                                  # definimos c
            MB[i,:] -= c*MB[j,:]                                 # definimos MB[i,:]
        end
    end
    
    x= zeros(n)                                                  # definimos x como el arreglo de ceros 
    for i = n:-1:1                                               # para i de n a -1 a 1
        x[i] = MB[i,n+1]                                         # definimos xi
       for j= i+1:n                                              # para j de i-1 a n
          x[i] -= MB[i,j]*x[j]                                   # definimos xi
        end
         x[i] /= MB[i,i]
    end
      return x                                                   # regresamos a x
end

reduccion (generic function with 1 method)

In [61]:
M = [3.0 4.0 7.0; 9.0 5.0 3.0; 2.0 5.0 9.0]
B = [1.0; 4.0; 6.0]
reduccion(M,B)

3-element Vector{Float64}:
 -3.200000000000003
  8.600000000000007
 -3.4000000000000035

In [64]:
A = [3 4 9; 9 5 3; 2 5 9]
b = [1; 2; 7]
reduccion(A,b)

3-element Vector{Float64}:
 -1.2571428571428573
  3.7428571428571438
 -1.1333333333333337

**Ejercicio** Crea una función `solución` con un argumento `M`, que puedes suponer que es una matriz aumentada que representa un sistema de ecuaciones algebráicas linealmente independientes, que integre las funciones `reducción` y `retropropagación` y devuelva un arreglo con el vector solución $\vec{x}$.

In [5]:
function solucion(M)                                               # definimos la funcion                                       
    m,n = size(M)                                                  # definimos el tamaño de la matriz
    for i = 1:m                                                    # para i de 1 a m
        pivot = abs(M[i,i])                                        # definimos pivote como el valor absoluto de la matriz de i a i
        pivot_renglon = i                                          # definimos el renglon pivote
        for k = i+1:m                                              # para k de i+1 a m
            if abs(M[k,i]) > pivot                                 # si el valor absoluto de la matriz de k a i es mayor al pivote
            pivot = abs(M[k,i])                                    # deinimos pivote como el valor absoluto de la matriz de k a i 
                pivot_renglon = k                                  #definimos el reglon pivote como k
            end 
        end
        M[i,:], M[pivot_renglon,:] = M[pivot_renglon,:], M[i,:] 
            
            for j = i+1:m                                          # para j de i+1 hasta m
                factor = M[j,i]/ M[i,i]                            # definimos el factor
                M[j,:] -= factor * M[i,:]                          # definimos M[j,:] como la resta AND del factor por la matriz
            end
      end
                x = zeros(n-1)                                    # definimos x como el arreglo de ceros de n-1
                x[n-1] = M[m,n] / M[m, n-1]                       # definimos x de n-1
     for i = m-1:-1
                    sum = M[i,n]                                  # para i de m-1 a -1 defnimos la sua de M_i,n
                    for j = i+1:m                                 # para j de i+1 a m
                        sum -= M[i,j]*x[j]                        # definimod la suma igual a la resat AND de M por X
                    end
               x[i] = sum / M[i,i]                                # definimos xi como la suma entre la matriz
                    end
                        return x                                  # regresamos a x
                    end       

solucion (generic function with 1 method)

In [9]:
M = [3.0 4.0 7.0 9.0; 9.0 5.0 3.0 8.0; 2.0 5.0 9.0 6.0]
solucion(M)

3-element Vector{Float64}:
 0.0
 0.0
 3.800000000000004

## Cálculo de matrices inversas

Una de las características más importantes de las matrices no singulares (esto es, matrices cuadradas con determinante no nulo) es que son invertibles. Es decir que, si $A\in M_{n\times n}(\mathbb{R})$ es tal que $\text{det}(A)\neq 0$, entonces existe $C\in M_{n\times n}(\mathbb{R})$ tal que

$$ AC = I_{n\times n} = CA,$$

donde $I_{n\times n}$ es la matriz identidad con $n$ renglones y $n$ columnas. Observemos que, si esto sucede, entonces de la ecuación $AC = I_{n\times n}$ tenemos que el producto del primer vector columna de $C$ con la matriz $A$ es igual al vector
v
$$ \begin{pmatrix} 1 \\ 0 \\ \vdots \\ 0 \end{pmatrix}. $$

Más generalmente, el producto del $j$-ésimo vector columna de $C$ con la matriz $A$ dará como resultado un vector con $1$ en la $j$-ésima entrada y $0$ en todas las demás.

A raíz de esta observación, podemos armar y solucionar $n$ sistemas lineales de ecuaciones distintos y juntar los vectores solución en una matriz, obteniendo así una matriz _inversa por la derecha_ de $A$. Más aún, se puede ver sencillamente con un poquito de álgebra que esta matriz inversa por la derecha debe ser _la_ inversa. En efecto: sea $A\in M_{n\times n}(\mathbb{R})$ una matriz no singular (invertible), $A^{-1}$ su inversa y $C$ una matriz inversa de $A$ por la derecha; entonces

$$ \begin{align*} C &= I_{n\times n} C \\ &= (A^{-1}A)C  \\ &= A^{-1}(AC) \\ &= A^{-1}I_{n\times n} \\ &= A^{-1}.\end{align*} $$

**Ejercicio** Crea una función `invertir` con un argumento `M`, que puedes suponer que es una matriz cuadrada no singular (invertible), que devuelva la matriz inversa de `M`.

Sugerencia: Utiliza la función `solución` que creaste anteriormente.

In [2]:
M = [2 4 5 ; 7 6 9; 3 7 8]
inv(M)

3×3 Matrix{Float64}:
 -1.66667   0.333333   0.666667
 -3.22222   0.111111   1.88889
  3.44444  -0.222222  -1.77778

In [8]:
function invertir(V)                               # definimos la funcion invertir
    n0=length(V[:,1])                              # n0 com el tamaño de la matriz
    d=zeros(n0,n0)                                 # d como el arreglo de ceros
    count=0                                        # iniciamos el conteo en 0
   for i in 1:n0                                   # para una i de 1 a n0
        for k in 1:n0                              # para k en 1 a n0
            D[i,k]=solucion(V)             
            # D como la solucion de la matriz aumentada
        end
        count=count+n0
    end
    return D                                       # obtenemos D como la salida
end

invertir (generic function with 1 method)

In [38]:
function invertir_matriz(M)
    m, n = size(M)
    I = Matrix{Float64}(I, m, m)
    for i in 1:m
        I[i, i] = 1.0
    end
    
    A = [M I]  # Concatenamos la matriz M con la matriz identidad para aplicar eliminación gaussiana
    
    solucion(A)  # Utilizamos la función solucion para resolver el sistema de ecuaciones
    
    inv_M = A[:, n+1:end]  # Obtenemos la submatriz correspondiente a la inversa de M
    
    return inv_M
end

invertir_matriz (generic function with 1 method)

In [39]:
M = [3.0 4.0 7.0; 9.0 5.0 3.0; 2.0 5.0 9.0]
I[i, i] = 1.0

invertir_matriz(M)

LoadError: UndefVarError: i not defined

In [6]:
#Función invertir
function invertir1(M)

c=size(M)[2]
f=size(M)[1] #Especificación número de columnas y filas de M

M′=zeros(f,c) #Matriz inversa primitiva (sólo ceros) 
    
    for j in 1:f
        y=zeros(3) #Vector ceros
        y[j]=1 #Vector con ceros y 1 en la j-ésima fila

        B=hcat(M,y) #Matriz para resolver ecuaciones lineales
        b=solucion(B) #Vector solución de matriz B

        for i in 1:c
            M′[i,j]=b[i] #Llenado matriz inversa
        end

    end
return M′
end

invertir1 (generic function with 1 method)

In [7]:
#Prueba de función
M=[1.0 2.0 2.0; 1.0 4.0 9.0; 1.0 3.0 6.0]
invertir1(M)

3×3 Matrix{Float64}:
  0.0   0.0  0.0
  0.0   0.0  0.0
 -1.0  -1.0  2.0

## El método de propagación

Existe otro método para solucionar sistemas lineales de ecuaciones algebráicas **totalmente análogo al que hemos visto**. En efecto, esto se sigue de observar que, si en vez de tener un sistema cuya matriz asociada sea triangular superior, fuese triangular _inferior_

$$ \begin{pmatrix} a'_{11} &0 &\dots &0 \\ a'_{21} &a'_{22} &\dots &0 \\ \vdots &\vdots &\ddots &\vdots \\ a'_{n1} &a'_{n2} &\dots &a'_{nn} \end{pmatrix} \begin{pmatrix} x_1 \\ x_2 \\ \vdots \\ x_n \end{pmatrix} = \begin{pmatrix} b'_1 \\ b'_2 \\ \vdots \\ b'_n \end{pmatrix} $$

entonces podríamos solucionarlo de principio a fin, obteniendo primero a

$$x_1 = \frac{b'_{1}}{a'_{11}}$$

y luego la fórmula general

$$ x_i = \frac{1}{a'_{ii}} \bigg( b'_i - \sum_{j=1}^{i-1} a'_{ij}x_j \bigg) $$

para $1<i\leq n$. A este método se le conoce como _propagación_, pues solucionamos la primera entrada del vector incógnita y después propagamos esa solución iterativamente hacia adelante. Asimismo, podemos reducir una matriz aumentada de la forma

$$ \begin{pmatrix} a_{11} &a_{12} &\dots &a_{1n} &\mid &b_1 \\ a_{21} &a_{22} &\dots &a_{2n} &\mid &b_2 \\ \vdots &\vdots &\ddots &\vdots &\mid &\vdots \\ a_{n1} &a_{n2} &\dots &a_{nn} &\mid &b_n \end{pmatrix} $$

a una de la forma

$$ \begin{pmatrix} a'_{11} &0 &\dots &0 &\mid &b'_1 \\ a'_{21} &a'_{22} &\dots &0 &\mid &b'_2 \\ \vdots &\vdots &\ddots &\vdots &\mid &\vdots \\ a'_{n1} &a'_{n2} &\dots &a'_{nn} &\mid &b'_n \end{pmatrix} $$

haciendo un proceso inverso al que vimos anteriormente -siempre y cuando, nuevamente, la matriz cuadrada sea no singular.

En inglés, este segundo método (propagación) se conoce como _forward substitution_, mientras que el primero que vimos (retropropagación) se conoce como _backward substitution_.

## Recursos complementarios

* Capítulo 6 "Direct Methods for Solving Linear Systems" de Burden, Faires y Burden, _Numerical Analysis_ (2016).
* Secciones 3.2 "Solution of Triangular Systems" y 3.3 "The Gaussian Elimination Method (GEM) and LU Factorization" de Quarteroni, Sacco y Saleri, _Numerical Mathematics_ (2000).
* Capítulo 2 "Systems of Linear Equations" de Poole, _Linear Algebra: A Modern Introduction_ (2015).